magento/pwa-studio
# HTML Caching and SW Communication API Implementation.

## Description

HTML Caching and Service Worker Communication API Implementation. The service worker will be following the `StaleWhileRevalidate` strategy for all HTML files. It will return the readily available file from the cache and try to update the file in the background. If the file happens to be different from the one already in the cache, it will send a message to the UI about an update available and a button in the message to refresh the page.

## Related Issue
Closes #1772 #1781 

### Verification Steps
1. Build and start the server using `yarn build && yarn stage:venia`.
2. Refresh the page once after it has loaded up. This will make sure that the Service Worker is installed successfully.
3. Open `venia-concept/dist/index.html` and make a change like adding a new script tag
`<script>console.log('Hey this is a new HTML')</script>`
4. Now navigate to a different page like Category or Product by clicking on the appropriate links/pictures. Note: Do not refresh the page.
5. Once you navigate to a new page, a toast should show up in the bottom letting you know that an update is available.

## Screenshots / Screen Captures (if appropriate)
![HTML Update Toast Demo](https://user-images.githubusercontent.com/35203638/67125172-cce03400-f1b9-11e9-9dfa-40af44dcbf75.gif)

## Checklist
* No functional change. Make sure both `yarn stage:venia` and `yarn watch:venia` should work like previous.
* Should be merged only after #1874 is merged.
